feature: use http caching layer in front of the github api

asvoboda opened the pull request:
The httpcache package is [recommended](https://github.com/google/go-github#conditional-requests) for the purpose of http caching by the go-github readme.

asvoboda committed:
feature: use http caching layer in front of the github api

asvoboda commented:
It seems like httpcache exposes 'hits' as a header: https://github.com/gregjones/httpcache/blob/master/httpcache.go#L157

I think ideally we would translate that into a metric, which could then be picked up by something like datadog.

asvoboda committed:
improvement: add caching to metrics registry

bluekeyes commented on githubapp/client_creator.go:
```diff
@@ -201,7 +202,11 @@ func (c *clientCreator) newClient(base *http.Client, details string) (*github.Cl
+	cachedTransport := httpcache.NewTransport(cache)
+	cachedTransport.Transport = base.Transport
+
+	client := github.NewClient(cachedTransport.Client())
```
Maybe this won't work, but I was thinking this would be optional middleware in the same way that client logging and metrics are. Since the cache transport is an implementation of `http.RoundTripper` it should be easy to integrate as middleware.

That would also allow users to configure the cache, which I think is a requirement.

asvoboda committed:
improvement: use ClientMiddleware to apply the http cache

asvoboda committed:
Revert "improvement: use ClientMiddleware to apply the http cache"

This reverts commit b41ce05d5909b860f210b1b814c6363af113a5a0.

asvoboda committed:
improvement: bring your own cache

bluekeyes commented on githubapp/client_creator.go:
```diff
@@ -132,6 +134,13 @@ func WithClientUserAgent(agent string) ClientOption {
 }
 
+// WithClientCache sets the http cache for all created clients
+func WithClientCache(cache httpcache.Cache) ClientOption {
```
After thinking about this more, I believe the user actually needs to provide a `func() httpcache.Cache` instead of a single instance. Cache keys appear to use only the URL and method, meaning they don't account for authentication. If we share a single cache between all clients, I think we could easily expose private repositories or other information to clients that don't have permissions to see them.

bluekeyes commented on githubapp/client_creator.go:
```diff
@@ -219,7 +235,14 @@ func (c *clientCreator) newV4Client(base *http.Client, details string) (*githubv
+		baseClient = cachedTransport.Client()
+	}
+
+	client := githubv4.NewEnterpriseClient(v4BaseURL.String(), baseClient)
```
I think using this for the GraphQL API is unnecessary. Only `GET` and `HEAD` requests are cached, and all GraphQL interaction happens via `POST` requests.

bluekeyes reviewed (commented):
While I'm not opposed to implementing this as a top-level parameter, I am curious: what didn't work about the middleware implementation?

asvoboda committed:
improvement: provide a func for client creators

ensure that each installation has its own unique cache
to prevent data leaks between installations

asvoboda committed:
docs: fix example

asvoboda commented:
re middleware vs top level approach: I was having trouble with the interaction between a `cachedTransport.Client()` and what we already had available in the middleware, which manifested in authentication errors. I suspect it might also be a case of handling it wrong.

asvoboda commented on githubapp/client_creator.go:
```diff
@@ -132,6 +134,13 @@ func WithClientUserAgent(agent string) ClientOption {
+// WithClientCache sets the http cache per constructed client
+func WithClientCache(cacheFunc interface{}) ClientOption {
+	return func(c *clientCreator) {
+		c.cacheFunc = cacheFunc.(func() *httpcache.Cache)
```
This part isn't yet working:
```
interface conversion: interface {} is func() *httpcache.MemoryCache, not func() *httpcache.Cache
```

bluekeyes replied:
I'd recommend specifying this signature as a `func() httpcache.Cache` and then wrapping the constructor in an anonymous function:

```go
WithClientCache(func() httpcache.Cache { return httpcache.NewMemoryCache() })
```

It's extra code, but I think it's the only way to satisfy the type system and still accept arbitrary implementations of `httpcache.Cache`.

bluekeyes commented:
I envisioned the middleware working like this (untested):

```go
func ClientCaching(cache func() httpcache.Cache) ClientMiddleware {
  return func(next http.RoundTripper) http.RoundTripper {
    cached := httpcache.NewTransport(cache())
    cached.Transport = next
    return cached
  }
}
```

If it works, that's still my preference, as caching seems more appropriate as middleware, like the logging and the metrics.

asvoboda committed:
fix: accept arbitrary httpcache types

asvoboda committed:
caching as middleware
